### PR TITLE
Revit2017 rooms

### PR DESCRIPTION
**Pull request by moethu**
### Purpose

Exposing Revit Room element to Dynamo.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] Snapshot of UI changes, if any.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/Libraries/RevitNodes/Elements/Room.cs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        internal Autodesk.Revit.DB.Architecture.Room InternalRevitElement
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// Reference to the Element
+        /// </summary>
+        public override Autodesk.Revit.DB.Element InternalElement
```
Comment (sharadkjaiswal) [inline src/Libraries/RevitNodes/Elements/Room.cs]: You need to put attribute [SupressImportIntoVM] on the public properties or methods that uses types from Autodesk.Revit.DB namespace. We don't want this type to be imported in DesignScript VM.

File: src/Libraries/RevitNodes/Elements/Room.cs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            }
+            else
+            {
+                // Update Location only
+                Autodesk.Revit.DB.LocationPoint point = (Autodesk.Revit.DB.LocationPoint)RoomElem.Location;
+                point.Point = location;
+            }
+
+            // Apply name and number if set
+            if (name != string.Empty)
```
Comment (sharadkjaiswal) [inline src/Libraries/RevitNodes/Elements/Room.cs]: better to use string.IsNullOrEmpty(name).

File: src/Libraries/RevitNodes/Elements/Room.cs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+        #endregion
+
+        #region Properties
+
+        /// <summary>
+        /// Get room name
+        /// </summary>
+        public string GetName
```
Comment (sharadkjaiswal) [inline src/Libraries/RevitNodes/Elements/Room.cs]: Property name should not include Get. Can we have Name, Number,Area,Height etc. as property?

File: src/Libraries/RevitNodesUI/GenericClasses.cs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+        protected override CoreNodeModels.DSDropDownBase.SelectionState PopulateItemsCore(string currentSelection)
+        {
+            PopulateItems();
+            return SelectionState.Done;
+        }
+        /// <summary>
+        /// Populate the Dropdown menu
+        /// </summary>
+        public void PopulateItems()
```
Comment (sharadkjaiswal) [inline src/Libraries/RevitNodesUI/GenericClasses.cs]: Base class already has an implementation of PopulateItems() which calls PopulateItemsCore() method. You can move the implementation of PopulateItems() to PopulateItemsCore() method.

File: src/Libraries/RevitNodesUI/GenericClasses.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        {
+            if (this.ElementType != null)
```
Comment (sharadkjaiswal) [inline src/Libraries/RevitNodesUI/GenericClasses.cs]: You may consider if(this.ElementType == null) return;

**Comment by moethu** [comment]
@sharadkjaiswal thanks for the comments - I haven't added the changes from the 2016 branch yet. Will do this now. They incorporate Michael's review.

About the [SupressImportIntoVM] Attribute - my understanding was that internal properties aren't displayed in Dynamo anyway, since none of the other elements have this attribute.

**Comment by moethu** [comment]
@sharadkjaiswal like this? https://github.com/DynamoDS/DynamoRevit/pull/1017/commits/eaa5aca4fa783a7237bb99d034c3f6503a1f7cbd

File: src/Libraries/RevitNodes/Elements/Room.cs [hunk 6]
```diff
@@ -21,6 +21,7 @@ public class Room : Element
         /// <summary>
         /// Internal reference to the Revit Element
         /// </summary>
+        [SupressImportIntoVM]
```
Comment (sharadkjaiswal) [inline src/Libraries/RevitNodes/Elements/Room.cs]: If the method, property or class is internal then you don't need this attribute. But if it's public then you do need to specify this attribute when you are using Autodesk.Revit.DB namespace type in your method/property.

**Comment by mjkkirschner** [comment]
@moethu has this been updated with changes from the 2016 branch?

**Comment by moethu** [comment]
@mjkkirschner I will check. let me cherry pick the latest commits from 2016 first. Is the 2016 version approved so far?

**Comment by mjkkirschner** [comment]
@moethu yes.

File: src/Libraries/RevitNodes/Elements/Room.cs (hunk 1 above)
Comment (mjkkirschner) [inline src/Libraries/RevitNodes/Elements/Room.cs]: this is still missing the Suppress import attribute.

**Comment by mjkkirschner** [comment]
@moethu there are some merge conflicts on this PR, and a missing suppression attribute noted above. Can you try to sync up 2016 and 2017 branches and then we can get this merged to avoid any more code rot.

**Comment by moethu** [comment]
@mjkkirschner applied all changes and had to rebase - same like 2016

**Comment by mjkkirschner** [comment]
github reports merge conflicts here :(

**Comment by moethu** [comment]
@mjkkirschner  resolved

**Pull request merged**